zqd/storage/zng.go

```diff
@@ -0,0 +1,148 @@
+package storage
+
+import (
+	"context"
+	"errors"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync/atomic"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio"
+	"github.com/brimsec/zq/zio/zngio"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zqe"
+)
+
+const (
+	allZngFile    = "all.zng"
+	allZngTmpFile = "all.zng.tmp"
+)
+
+type ZngStorage struct {
+	path       string
+	writing    uint32
+	span       nano.Span
+	index      *zngio.TimeIndex
+	streamsize int
+}
+
+func NewZng(path string) *ZngStorage {
+	return &ZngStorage{path: path, index: zngio.NewTimeIndex()}
+}
+
+func (s *ZngStorage) join(args ...string) string {
+	args = append([]string{s.path}, args...)
+	return filepath.Join(args...)
+}
+
+func (s *ZngStorage) Open(span nano.Span) (zbuf.ReadCloser, error) {
+	zctx := resolver.NewContext()
+	f, err := os.Open(s.join(allZngFile))
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+
+		// Couldn't read all.zng, check for an old space with all.bzng
+		bzngFile := strings.TrimSuffix(allZngFile, filepath.Ext(allZngFile)) + ".bzng"
+		f, err = os.Open(s.join(bzngFile))
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return nil, err
+			}
+			r := zngio.NewReader(strings.NewReader(""), zctx)
+			return zbuf.NopReadCloser(r), nil
+		}
+	}
+	// XXX plz keep track of how long this takes
+	return s.index.NewReader(f, zctx, span)
+}
+
+type spanWriter struct {
+	span nano.Span
+}
+
+func (w *spanWriter) Write(rec *zng.Record) error {
+	if rec.Ts == 0 {
+		return nil
+	}
+	first := w.span == nano.Span{}
+	s := nano.Span{Ts: rec.Ts, Dur: 1}
+	if first {
+		w.span = s
+	} else {
+		w.span = w.span.Union(s)
+	}
+	return nil
+}
+
+// XXX only one writeto allowed at a time
+func (s *ZngStorage) WriteTo(ctx context.Context, zr zbuf.Reader) error {
```

Comment:
As I suggested previously, renaming this to "Replace" would better communicate what it currently does.

Reply:
Yep.